### DISCUSSION TITLE
Best practice for "external libraries" and cmake

### DISCUSSION THREAD

**Discussion by boaks** [zephyrproject-rtos/zephyr#43650] [Q&A]
Is there a way of "best practice" to prepare cmake files of an "external library".
Especially, if such a library comes already with cmake files or is intended to support other OS as well.

I found also a discussion in issue #8439 but that doesn't include such best practice.

**Comment by nashif** [comment]
@tejlmand  FYI

**Comment by boaks** [comment]
Thanks for moving. I overseen that "Discussions" are available.

**Comment by yashi** [comment]
I have been also searching for a best practice.  I've written what I learnt while back here https://github.com/yashi/module-sample.  I've posted it in discord thread here https://discord.com/channels/720317445772017664/927715693636517958.

**Reply by boaks** [reply]
I will have a look at it, thanks.

**Reply by boaks** [reply]
Maybe I'm too old fashion, but I can't see you "discord channel". Is there any special to be done to have access?

**Reply by yashi** [reply]
can you access https://chat.zephyrproject.org/ ?

**Reply by yashi** [reply]
If not, don't worry.  The thread doesn't have much.

**Reply by boaks** [reply]
Yes, I can access it! Thanks a lot. 
My forecast for this week is a little "over-busy" with [eclipse/californium](https://github.com/eclipse/californium) in order to prepare for 3.4 there. But after that, I will have a look at it. I'm really interested, to publish a port of [eclipse/tinydtls - cid](https://github.com/eclipse/tinydtls/tree/feature/connection_id) for zephyr, even if that will take a while ;-).

**Reply by boaks** [reply]
I started to prepare my work according to your example.
Unfortunately, I was still busy with other jobs, so I could just start and I'm not ready. I will update this, if I'm done.
Thanks anyway for the example, that helps a lot to understand, how it is intended to be done.

**Reply by boaks** [reply]
I'm some steps further.

One difference, which seems for me to be not too easy to overcome is, that some (small) parts of my module-library depends on 

```
#include <zephyr.h> 
```

I searched, I read a lot, but I didn't find, what is required for the library target to add the zephyr include path.

[external_lib/CMakeLists.txt](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/application_development/external_lib/CMakeLists.txt) that is then using a different build system. For just add the zephyr-include-path to the libraries, that seems to be a overkill.

So, is there a way to extend:

```
target_include_directories(tinydtls PUBLIC . aes ecc sha2)
```

to add the zephyr include-path? 

(I guess, the next question will then be, how to pass also the zephyr compiler definitions to 

```
target_compile_definitions(tinydtls PUBLIC ...)
```
)

**Reply by boaks** [reply]
```
target_link_libraries(tinydtls PUBLIC zephyr_interface)
```

Was the missing link ...

Now I need to try again to split the stuff as in [yashi/module-sample](https://github.com/yashi/module-sample).

**Reply by yashi** [reply]
Let me know if there is anything I should fix.

**Reply by boaks** [reply]
I managed to apply the pattern in [yashi/module-sample](https://github.com/yashi/module-sample) to tinydtls. It compiles and can run it on the target.

@yashi  thanks a lot for your great example.
